expanding table rows

cforkish opened the pull request:
this PR adds the ability to expand a student row in our section progress tables to display two additional detail rows: time spent, and last updated timestamp.

<img width="582" alt="Screen Shot 2021-03-17 at 4 21 13 PM" src="https://user-images.githubusercontent.com/4986878/111553673-1bdd2080-8742-11eb-8245-1610a3ded7f6.png">
<img width="583" alt="Screen Shot 2021-03-17 at 4 21 29 PM" src="https://user-images.githubusercontent.com/4986878/111553675-1c75b700-8742-11eb-9347-09868243b67c.png">

Note: there is one more update in progress for the detail view table that needs to be merged into this branch before this PR can be merged. i will add screenshots of the detail table once that work is complete.

## Links
- [spec](https://docs.google.com/document/d/11vid6ZWkpUV5TJm3swI6kbdSpg5_93UHDxnod58r_Zc/edit#bookmark=id.hnhfxddo3gre)
- [mock](https://www.figma.com/proto/Rc0o2mLaPx3gjI7WKSU6CBBH/CSF-Progress-Reports?node-id=447%3A902&scaling=min-zoom)
- [jira epic](https://codedotorg.atlassian.net/browse/LP-1451)


## Deployment strategy
this new functionality is hidden behind an experiment/pilot flag, so it will only show up for users who are enrolled in the pilot.

maureensturgeon commented on apps/src/templates/sectionProgress/progressTables/ProgressTableContainer.jsx:
```diff
@@ -80,27 +103,109 @@ class ProgressTableContainer extends React.Component {
+      return row.student.id !== rowData.student.id || row.expansionIndex === 0;
+    });
+    rows[rowIndex].isExpanded = false;
+    this.setState({
```
Fun fact: this could also be written as `this.setState({rows})` since the key and value have the same name (also on line 152)

cforkish replied:
👍

maureensturgeon commented on apps/src/templates/sectionProgress/progressTables/ProgressTableContainer.jsx:
```diff
@@ -80,27 +103,109 @@ class ProgressTableContainer extends React.Component {
+            columnWidths={this.props.columnWidths}
+            lessonCellFormatters={this.props.lessonCellFormatters}
+            extraHeaderFormatters={this.props.extraHeaderFormatters}
+            includeHeaderArrows={this.props.includeHeaderArrows}
```
On the next line it looks like you're still passing all props through, do you still need all those?

cforkish replied:
nope!

maureensturgeon commented on apps/test/unit/templates/sectionProgress/progressTables/ProgressTableStudentListTest.jsx:
```diff
@@ -48,13 +55,27 @@ describe('ProgressTableStudentList', () => {
+    );
+    expect(wrapper.find(ProgressTableTextLabelCell)).to.have.lengthOf(2);
+
+    sinon.restore();
```
I thought that you needed to call `restore()` on the stub, not sinon (that's what it looks like in the docs). Are you sure this is the right syntax?

cforkish replied:
`sinon` is an alias for the default [sandbox](https://sinonjs.org/releases/v9.2.4/sandbox/), so calling `sinon.restore()` restores the whole sandbox.

jamescodeorg commented on apps/src/templates/sectionProgress/progressTables/ProgressTableLevelIconSet.jsx:
```diff
@@ -0,0 +1,73 @@
+  </div>
+);
+
+function LevelIcon({icon, isUnplugged}) {
```
A header comment here describing what this component does and/or how it is used would be helpful.

cforkish replied:
ugh. this file is not new, but github isn't picking up the rename. when i realized that the detail row text required the exact same layout as the level icons, i factored out that layout logic into `ProgressTableLevelSpacer`, and in the process kept getting confused about the fact that this component displays icons for multiple levels, so i renamed it. you can view the refactor by filtering the PR to all the commits prior to the rename commit.

that said, i should still add header comments!

jamescodeorg commented on apps/src/templates/sectionProgress/progressTables/ProgressTableLevelSpacer.jsx:
```diff
@@ -0,0 +1,56 @@
+  sublevelCount: PropTypes.number.isRequired
+};
+
+export default function ProgressTableLevelSpacer({items}) {
```
A header comment here describing what this component does and/or how it is used would be helpful.

Is the primary purpose of this component to set the width of the cells?  Or maybe to ensure that bubbles are spaced evenly in each cell?

cforkish replied:
added a comment, let me know if it clarifies

jamescodeorg replied:
The new header comments are excellent, thanks!

jamescodeorg reviewed (approved):
I think I understand how all the pieces of this fit together, but this feels like this is introducing a fair amount of complexity that might be tricky for someone to understand in the future.  Maybe a little bit of extra commenting and updating the global design doc would be helpful?

cforkish commented:
@jamescodeorg your comment about complexity is a bit disappointing to hear because i was pleasantly surprised to find that this change was as simple as it was (from my point of view). i'm hoping that the total lack of comments was the primary culprit of it seeming complex, plus maybe the fact that github is making it look like i added a whole new component for level icons (when in fact i simplified the existing one). hopefully the comments i added to `ProgressTableContainer` will go a long way toward clarifying things. apologies for not having those in place while you were trying to figure out how it all works!
